lib/ansible/modules/cloud/docker/docker_swarm_facts.py

```diff
@@ -0,0 +1,319 @@
+#!/usr/bin/python
+#
+# (c) 2019 Piotr Wojciechowski <piotr@it-playground.pl>
+# (c) Thierry Bouvet (@tbouvet)
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: docker_swarm_facts
+
+short_description: Retrieves facts about Docker Swarm cluster.
+
+description:
+  - Retrieves facts about a Docker Swarm.
+  - Returns lists of swarm objects names for the services - nodes, services, tasks.
+  - The output differs depending on API version available on docker host.
+  - Must be run on Swarm Manager node.
```

Comment:
I think the module should also return something sensible (which is not an error) when it is not a manager node, or the docker daemon is not in swarm mode at all. Or, if that makes implementation too complicated, `docker_host_facts` should return something like that :-)

Reply:
Why not fail? `docker_node_facts` fails in cases it is not run on Swarm Manager. It may be easier to handle conditional in playbook if it fails

Reply:
The main reason is that usually, `xxx_facts` modules also return information on whether `xxx` is available at all. It would be fine for me to simply return more detailed error messages on failure, after all we can pass arbitrary information along when calling `module.fail_json()`. If we return this information (docker in swarm mode / node is manager) also on failure situations, users can add `ignore_errors: yes` and use that information if they like.

Reply:
Ok, will extend the code in such way

Reply:
Are you still working on this? I can't find anything like that in the current version.

Reply:
Line `168` calls the `fail_task_if_not_swarm_manager()`, I also updated the error message in the library to be more informative and consistent so ever swarm module fails the same way if not run on a manager

Reply:
Ok, but that's only the message, which shouldn't be parsed / understood by the calling code. Something useable would be a boolean variable `docker_in_swarm_mode` or `swarm_node_is_manager` which is returned in any case (except maybe when the docker daemon doesn't seem to be there in the first place).

It might be a good idea to introduce a general `fail_results` variable in `AnsibleDockerClient`, which is empty by default and passed on to `module.fail_json` in `fail()`. If everyone (i.e. also the code in module_utils/docker/swarm.py) calls `client.fail()` instead of `module.fail_json()`, it will always be returned.

Then the facts module can put the booleans into `client.fail_results` with value `False`, and update them to `True` once the corresponding calls succeed.

If you want I can prepare such a system in another PR.

Reply:
That is an interesting idea and can be later implemented in `fail_task_if_not_swarm_manager()` and other `AnsibleDockerSwarmClient`. In such a case, this new feature should not block this PR.

Reply:
Fine for me. 2.8 is still some some time out, fortunately :)

Reply:
See #51999 for the promised PR. (I also added some info on the 2.8 timeline to the agenda/task list; this probably interests everyone and shouldn't be buried in a comment here ;) )

Reply:
Good :) If nobody else has anything to add let's put ansibot to work to set required labels to merge it :)